### ISSUE TITLE
Training classifiers with natural language explanations

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#366]
Training classifiers with natural language explanations<br>
<br>
https://ift.tt/2PzKrhF<br>
<br>
adriancolyer<br>
<br>
<p><a href="https://arxiv.org/abs/1805.03818">Training classifiers with natural language explanations</a> Hancock et al., <em>ACL&rsquo;18</em></p>
<p>We looked at <a href="https://blog.acolyer.org/2018/08/22/snorkel-rapid-training-data-creation-with-weak-supervision">Snorkel</a> earlier this week, which demonstrates that maybe AI isn&rsquo;t going to take over all of our programming jobs. Instead, we&rsquo;ll be writing <em>labelling functions</em> to feed the machine! Perhaps we could call this task <em>label engineering</em>. To me, it feels a bit like programming a quick-and-dirty expert system, where the downstream generative model deals with all the inaccuracies and inconsistencies so that we don&rsquo;t have to be perfect, just useful. Given the success of the approach, a natural question to ask is how we can enable end users to more easily create useful labelling functions in their domain. This is where <em>BabbleLabble</em> comes in!</p>
<blockquote>
<p>In this work, we propose BabbleLabble, a framework for training classifiers in which an annotator provides a natural language explanation for each labeling decision. A semantic parser converts these explanations into programmatic labeling functions that generate noisy labels for an arbitrary amount of unlabeled data, which is used to train a classifier.</p>
</blockquote>
<p>So much for those programming jobs ! <img alt="&#128521;" src="https://s0.wp.com/wp-content/mu-plugins/wpcom-smileys/twemoji/2/72x72/1f609.png" class="wp-smiley"></p>
<p>Working with BabbleLabble, it takes users about twice as long per example to provide a label <em>plus an explanation</em> as it does just to label an example. (The user has to do the work of figuring out the rationale in both cases). By automatically distilling a set of labelling functions from the explanations though (we&rsquo;ll get to how that bit works shortly), the end result is that on three tasks <em>users were able to train classifiers with comparable F1 scores from 5-100x faster</em>.</p>
<p>If you combine BabbleLabble with Snorkel and the advances in machine learning models, you can start to see a new end-to-end approach for the rapid creation of intelligent systems. It starts with <em>domain knowledge capture</em> systems, of which simply labelling an example by hand is the simplest example. Snorkel and BabbleLabble show us that we can do so much more here to capture and exploit available domain knowledge. Domain knowledge capture systems assist in the task of <em>label engineering</em>, and their output is a set of labelling functions. A <em>probabilistic label generator</em> intelligently combines the labelling functions to create a set of probabilistic labels (<em>data programming</em>), and finally we bolt on the end the supervised machine learning model of our choice and train it using the generated labels.</p>
<p>What ties all this together is that via the wonders of data programming, the labelling functions don&rsquo;t have to be perfect. It&rsquo;s a much lower bar to write/generate a <em>useful</em> labelling function than it is to create one with very high accuracy and coverage.</p>
<h3>Don&rsquo;t lose the rationale</h3>
<p>When we ask a human annotator to provide a label for an example, we up with limited information (one bit, in the case of binary classification).</p>
<blockquote>
<p>This invites the question: how can we get more information per example, given that the annotator has already spent the effort reading and understanding an example? &hellip; In this work, we tap into the power of natural language and allow annotators to provide supervision to a classifier via <em>natural language explanations</em>.</p>
</blockquote>
<p>For example:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-1.jpeg?w=480"></p>
<p>The captured explanations are turned into candidate labeling functions, which go through a filtering process to produce a final set of labeling functions used to drive Snorkel.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-2.jpeg?w=640"><br>
(<a href="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-2.jpeg">Enlarge</a>)</p>
<p>What&rsquo;s very neat is that the semantic parser used to turn explanations into candidate labeling functions is generic &ndash; i.e., it the same parser can be used across different tasks without any task-specific training.</p>
<blockquote>
<p>One of our major findings is that in our setting, even a simple rule-based parser suffices&hellip;</p>
</blockquote>
<p>This works because the filtering stage removes inconsistent and unhelpful (e.g., assigning the same label to every example) labelling functions, and Snorkel&rsquo;s data programming copes with inconsistencies and inaccuracies in those that remain. Another very interesting finding is that labelling functions that are <em>near</em> to the perfect logical capture of the explanation often perform better than a perfect capture. In the evaluation, this is explained thus:</p>
<blockquote>
<p>In other words, when users provide explanations, the signals they describe provide good starting points, but they are actually unlikely to be optimal.</p>
</blockquote>
<h3>Generating labeling functions</h3>
<p>The first step in the process is the use of a semantic parser to parse explanations and generate candidate labelling functions. Explanations follow a loose grammar (see example predicates in appendix A of the paper), and the parser iterates over input subspans looking for rule matches:</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-3.jpeg?w=640"></p>
<blockquote>
<p>We emphasize that the goal of this semantic parser is <em>not</em> to generate the single correct parse, but rather to have coverage over many potentially useful LFs.</p>
</blockquote>
<p>Users of the system are provided with minimal examples of all of the predicates in the grammar in use. The construction of the grammar was guided by analysing 500 free-form user-provided explanations on a <em>spouse</em> task. The resulting predicates are summarised in the table below. There are 200 rule templates in total.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-table-1.jpeg?w=480"></p>
<h3>Distilling labeling functions</h3>
<p>The set of candidate labelling functions (LFs) produced by parsing an explanation are then filtered. A <em>semantic filter</em> checks that a candidate LF is actually consistent with its own example.</p>
<p>For example, given an explanation &ldquo;True, because the words &lsquo;his wife&rsquo; are right before person 2&rdquo;, there are two possible interpretations of &lsquo;right&rsquo;. It could be &lsquo;right before&rsquo; or &lsquo;to the right of&rsquo;. A candidate LF generated from the latter interpretation can be excluded when the example in question is &ldquo;Tom Brady and his wife Gisele Bundchen were spotted in New York City on Monday amid rumors&hellip;&rdquo;. (Because the words &lsquo;his wife&rsquo; are actually to the <em>left</em> of person 2).</p>
<p>A set of <em>pragmatic filters</em> remove LFs that are constant (label every example the same), redundant (labels the training set identically to some other LF), or correlated.</p>
<blockquote>
<p>Finally, out of all LFs from the same explanation that pass all the other filters, we keep only the most specific (lowest coverage) LF. This prevents multiple correlated LFs from a single example dominating.</p>
</blockquote>
<p>Over the three tasks in the evaluation, the filter bank removes over 95% of incorrect parses, and the incorrect ones that remain have average end-task accuracy within 2.5 points of the corresponding correct parses.</p>
<h3>Experimental results</h3>
<p>We already know from <a href="https://blog.acolyer.org/2018/08/22/snorkel-rapid-training-data-creation-with-weak-supervision">Snorkel</a> what happens downstream of the labelling function selection process, so let&rsquo;s jump straight to the evaluation section of the paper. BabbleLabble is used on three tasks, <em>Spouse</em>, <em>Disease</em>, and <em>Protein</em>.</p>
<p>In the spouse task annotators see a sentence with two highlighted names and are asked to label whether the two people are spouses (sentences taken from the Signal Media news articles dataset).</p>
<p>In the disease task annotators are shown a sentence with highlighted names of a chemical and a disease, and asked whether the sentences suggests the chemical causes the disease. Because the necessary domain expertise was not readily available, explanation sentences in this case were created by having someone unfamiliar with BabbleLabble translate from the labelling functions generated as part of the <a href="https://blog.acolyer.org/2018/08/22/snorkel-rapid-training-data-creation-with-weak-supervision">Snorkel</a> publication evaluation.</p>
<p>The protein task was completed in conjunction with a neuroscience company called OccamzRazor. Annotators were shown a sentence from biomedical literature with highlighted names of a protein and kinase, and asked to label whether or not the kinase influences the protein in terms of a physical interaction or phosphorylation.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-table-2.jpeg?w=480"></p>
<p>Example sentences and explanations from each of the three tasks are shown in the figure below.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-4.jpeg?w=640"><br>
(<a href="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-4.jpeg">Enlarge</a>)</p>
<p>On an F1 score basis, the a classifier trained with BabbleLabble achieves the same accuracy as a classifier trained with just end labels, using between 5x to 100x fewer examples. On the spouse task, 30 explanations are worth somewhere around 5,000 labels, on the disease task 30 explanations are worth around 1,500 labels, and on the protein task 30 explanations are worth around 175 labels.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-table-3.jpeg?w=640"></p>
<blockquote>
<p>Once the number of labeled examples is sufficiently large, traditional supervision once again dominates, since ground truth labels are preferable to noisy ones generated by labeling functions. However, in domains where there is much more unlabeled data available than labeled data (which in our experience is most domains), we can gain in supervision efficiency from using BabbleLabble.</p>
</blockquote>
<p>The team compared the accuracy of BabbleLabble <em>without</em> the LF filtering process (i.e., using <em>all</em> candidate LFs), BabbleLabble a as is, and BabbleLabble using &lsquo;perfect&rsquo; parsers generated by hand to accurately translate the explanation sentences. The results show that the simple rule-based semantic parser and the perfect parser have nearly identical average F1 scores.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-table-5.jpeg?w=480"></p>
<p>Incorrect LFs still often provide useful signal.</p>
<p><img src="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-5.jpeg?w=640"><br>
(<a href="https://adriancolyer.files.wordpress.com/2018/08/babblelabble-fig-5.jpeg">Enlarge</a>)</p>
<h3>Learn more</h3>
<ul>
<li>
<a href="https://hazyresearch.github.io/snorkel/blog/babble_labble.html">Babble Labble blog post</a> from the authors</li>
<li>
<a href="https://worksheets.codalab.org/worksheets/0x900e7e41deaa4ec5b2fe41dc50594548/">Code, data, and experiments</a> from this paper</li>
</ul>
<br>
<br>
<br>
<br>
BigData,AI|ML|DM|IR|PR<br>
<br>
via the morning paper https://ift.tt/1WJ1E4W<br>
<br>
August 24, 2018 at 01:00PM
